cmd/dev.go

```diff
@@ -0,0 +1,677 @@
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		logger.Println("proxy listening on", addr)
+		err := server.ListenAndServe()
+		if err != nil && err != http.ErrServerClosed {
+			logger.Println("proxy error", err)
+		}
+	}()
+
+	go func() {
+		sigs := make(chan os.Signal, 1)
+		signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+		<-sigs
+		logger.Println("shutting down all commands")
+		for _, command := range commands {
+			command.Process.Signal(syscall.SIGTERM)
+		}
+		server.Shutdown(context.Background())
+	}()
+
+	if open, _ := cmd.Flags().GetBool("open"); open {
+		browser.OpenURL(fmt.Sprintf("http://%s", addr))
+	}
+
+	wg.Wait()
+
+	// Wait a bit for all logs to be printed
+	time.Sleep(1 * time.Second)
+
+	return nil
+}
+
+func writePortFile(filepath string, port int) error {
+	portDir := path.Dir(filepath)
+	if _, err := os.Stat(portDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(portDir, 0755); err != nil {
+			return err
+		}
+	}
+
+	return os.WriteFile(filepath, []byte(fmt.Sprintf("%d", port)), 0644)
+}
+
+func proxyFromDir(micros []*shared.Micro, routeDir string) (*proxy.ReverseProxy, error) {
+	routes := make([]proxy.ProxyRoute, 0)
+	for _, micro := range micros {
+		portFile := path.Join(routeDir, fmt.Sprintf("%s.port", micro.Name))
+		if _, err := os.Stat(portFile); err != nil {
+			continue
+		}
+
+		microPort, err := parsePort(portFile)
+		if err != nil {
+			continue
+		}
+
+		target, _ := url.Parse(fmt.Sprintf("http://localhost:%d", microPort))
+
+		logger.Println("proxying", micro.Prefix(), "to", target.String())
+		routes = append(routes, proxy.ProxyRoute{
+			Prefix: micro.Prefix(),
+			Target: target,
+		})
+	}
+
+	return proxy.NewReverseProxy(routes), nil
+}
+
+func parsePort(portFile string) (int, error) {
+	// check if the port is already in use
+	portStr, err := os.ReadFile(portFile)
+	if err != nil {
+		return 0, err
+	}
+
+	return strconv.Atoi(string(portStr))
+}
+
+func isPortActive(port int) bool {
+	conn, err := net.Dial("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return false
+	}
+
+	conn.Close()
+	return true
+}
+
+func microCommand(micro *shared.Micro, command string, directory, projectKey string, port int) (*exec.Cmd, error) {
+	var devCommand string
+	if command != "" {
+		devCommand = command
+	} else if micro.Dev != "" {
+		devCommand = micro.Dev
+	} else if engineToDevCommand[micro.Engine] != "" {
+		devCommand = engineToDevCommand[micro.Engine]
+	} else {
+		return nil, fmt.Errorf("no dev command found for micro %s", micro.Name)
+	}
+
+	commandDir := directory
+	if micro.Src != "" {
+		commandDir = path.Join(directory, micro.Src)
+	} else {
+		commandDir = path.Join(directory, micro.Name)
+	}
+
+	environ := map[string]string{
+		"PORT":                      fmt.Sprintf("%d", port),
+		"DETA_PROJECT_KEY":          projectKey,
+		"DETA_SPACE_APP_HOSTNAME":   fmt.Sprintf("localhost:%d", port),
+		"DETA_SPACE_APP_MICRO_NAME": micro.Name,
+		"DETA_SPACE_APP_MICRO_TYPE": micro.Type(),
+	}
```

Comment:
In the Space runtime we also set the env var `DETA_SPACE_APP_VERSION` to the version of the release the instance is using ([Space Docs](https://deta.space/docs/en/basics/micros#pre-set-variables)). If the Micro expects the variable to be set it could break something, maybe we should set the env var to something like `dev`?

Might be good if the Micro can check if it is running under `space dev`, we could add another env var for this: `DETA_SPACE_DEV=true`. Since it is not running with the Space runtime we should also set `DETA_SPACE_APP` to `"false"`.

Reply:
Needs further spec on those, because we are not sure about what we want in `DETA_SPACE_APP_VERSION`, `DETA_SPACE_APP`.